src/introspection.rs

```diff
@@ -0,0 +1,323 @@
+            } => {
+                debug!("overwriting previous path {}", prev.display());
+
+                AddInterfaceError::InterfaceFile { path, backtrace }
+            }
+        }
+    }
+}
+
+/// Trait that permits a client to add and remove interfaces dynamically after being connected.
+#[async_trait]
+pub trait DynamicIntrospection {
+    /// Add a new [`Interface`] to the device interfaces.
+    async fn add_interface(&self, interface: Interface) -> Result<(), Error>;
+
+    /// Add one ore more [`Interface`] to the device introspection.
+    async fn extend_interfaces<I>(&self, interfaces: I) -> Result<(), Error>
+    where
+        I: IntoIterator<Item = Interface> + Send;
+
+    /// Add a new interface from the provided file.
+    async fn add_interface_from_file<P>(&self, file_path: P) -> Result<(), Error>
+    where
+        P: AsRef<Path> + Send;
+
+    /// Add a new interface from a string. The string should contain a valid json formatted
+    /// interface.
+    async fn add_interface_from_str(&self, json_str: &str) -> Result<(), Error>;
+
+    /// Remove the interface with the name specified as argument.
+    async fn remove_interface(&self, interface_name: &str) -> Result<(), Error>;
+}
+
+#[async_trait]
+impl<S, C> DynamicIntrospection for AstarteDeviceSdk<S, C>
+where
+    C: Register + Send + Sync,
+    S: Send + PropertyStore,
+{
+    async fn add_interface(&self, interface: Interface) -> Result<(), Error> {
+        // Lock for writing for the whole scope, even the checks
+        let mut interfaces = self.interfaces.write().await;
+
+        let map_err = interfaces
+            .validate(interface)
+            .map_err(AddInterfaceError::Interface)?;
+
+        let Some(to_add) = map_err else {
+            debug!("interfaces already present");
+
+            return Ok(());
+        };
+
+        self.connection.add_interface(&interfaces, &to_add).await?;
+
+        interfaces.add(to_add);
+
+        Ok(())
+    }
+
+    async fn extend_interfaces<I>(&self, added: I) -> Result<(), Error>
+    where
+        I: IntoIterator<Item = Interface> + Send,
+    {
+        // Lock for writing for the whole scope, even the checks
+        let mut interfaces = self.interfaces.write().await;
+
+        let to_add = interfaces
+            .validate_many(added)
+            .map_err(AddInterfaceError::Interface)?;
+
+        if to_add.is_empty() {
+            debug!("All interfaces already present");
+            return Ok(());
+        }
+
+        debug!("Adding {} interfaces", to_add.len());
+
+        self.connection
+            .extend_interfaces(&interfaces, &to_add)
+            .await?;
+
+        interfaces.extend(to_add);
+
+        debug!("Interfaces added");
+
+        Ok(())
+    }
+
+    async fn add_interface_from_file<P>(&self, file_path: P) -> Result<(), Error>
+    where
+        P: AsRef<Path> + Send,
+    {
+        let interface = fs::read_to_string(&file_path).map_err(|err| AddInterfaceError::Io {
```

Comment:
why didn't you use `tokio::fs` instead of `std::fs`?

Reply:
Also, there is another place where `std::fs` is used, you could fix also that one

Reply:
Because the existing function call was synchronous and there was no need to change the API. Changing this should be done in another issue, with its own considerations.

The builder is used at the start of the program, so blocking the thread should be not be a problem, and having a sync call is more ergonomic without all the awaits. Maybe we should provide both APIs. Either way, I'm opening a new issue about this.